File: research/query_service/ir/runtime/src/graph/element/vertex.rs
//
//! Copyright 2020 Alibaba Group Holding Limited.
//!
//! Licensed under the Apache License, Version 2.0 (the "License");
//! you may not use this file except in compliance with the License.
//! You may obtain a copy of the License at
//!
//! http://www.apache.org/licenses/LICENSE-2.0
//!
//! Unless required by applicable law or agreed to in writing, software
//! distributed under the License is distributed on an "AS IS" BASIS,
//! WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//! See the License for the specific language governing permissions and
//! limitations under the License.

use std::cmp::Ordering;
use std::hash::{Hash, Hasher};
use std::io;

use dyn_type::BorrowObject;
use ir_common::NameOrId;
use pegasus_common::codec::{Decode, Encode, ReadExt, WriteExt};

use crate::expr::eval::Context;
use crate::graph::element::{Element, GraphElement};
use crate::graph::property::DynDetails;
use crate::graph::{read_id, write_id, ID};

#[derive(Clone, Debug)]
pub struct Vertex {
    id: ID,
    label: Option<NameOrId>,
    details: DynDetails,
}

impl Vertex {
    pub fn new(id: ID, label: Option<NameOrId>, details: DynDetails) -> Self {
        Vertex { id, label, details }
    }
}

impl Element for Vertex {
    fn details(&self) -> Option<&DynDetails> {
        Some(&self.details)
    }

    fn as_graph_element(&self) -> Option<&dyn GraphElement> {
        Some(self)
    }

    fn as_borrow_object(&self) -> BorrowObject {
        self.id().into()
    }
}

impl GraphElement for Vertex {
    fn id(&self) -> ID {
        self.id
    }

    fn label(&self) -> Option<&NameOrId> {
        self.label.as_ref()
    }

    fn len(&self) -> usize {
        0
    }
}

impl Encode for Vertex {
    fn write_to<W: WriteExt>(&self, writer: &mut W) -> io::Result<()> {
        write_id(writer, self.id)?;
        self.label.write_to(writer)?;
        self.details.write_to(writer)?;
        Ok(())
    }
}

impl Decode for Vertex {
    fn read_from<R: ReadExt>(reader: &mut R) -> io::Result<Self> {
        let id = read_id(reader)?;
        let label = <Option<NameOrId>>::read_from(reader)?;
        let details = <DynDetails>::read_from(reader)?;
        Ok(Vertex { id, label, details })
    }
}

impl Context<Vertex> for Vertex {
    fn get(&self, _tag: Option<&NameOrId>) -> Option<&Vertex> {
        Some(&self)
    }
}

impl Hash for Vertex {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.id.hash(state)
    }
}

impl PartialEq for Vertex {
    fn eq(&self, other: &Self) -> bool {
        self.id() == other.id()
    }
}

impl PartialOrd for Vertex {
    // TODO: not sure if it is reasonable. Vertex may be not comparable.
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        self.as_borrow_object()
            .partial_cmp(&other.as_borrow_object())
    }
}

File: flex/engines/graph_db/runtime/execute/ops/update/vertex.cc
/** Copyright 2020 Alibaba Group Holding Limited.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 	http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "flex/engines/graph_db/runtime/execute/ops/update/vertex.h"
#include "flex/engines/graph_db/runtime/common/operators/update/get_v.h"
#include "flex/engines/graph_db/runtime/utils/expr_impl.h"
#include "flex/engines/graph_db/runtime/utils/params.h"
#include "flex/engines/graph_db/runtime/utils/utils.h"

namespace gs {
namespace runtime {
namespace ops {

class UGetVFromEdgeOpr : public IUpdateOperator {
 public:
  UGetVFromEdgeOpr(const GetVParams& params) : params_(params) {}
  ~UGetVFromEdgeOpr() = default;

  std::string get_operator_name() const override { return "UGetVFromEdgeOpr"; }

  bl::result<Context> Eval(GraphUpdateInterface& graph,
                           const std::map<std::string, std::string>& params,
                           Context&& ctx, OprTimer& timer) override {
    return UGetV::get_vertex_from_edge(
        graph, std::move(ctx), params_,
        [&](size_t idx, label_t label, vid_t vid) { return true; });
  }

 private:
  GetVParams params_;
};

class UGetVFromEdgeWithPredOpr : public IUpdateOperator {
 public:
  UGetVFromEdgeWithPredOpr(const GetVParams& params,
                           const common::Expression& expr)
      : params_(params), expr_(expr) {}
  ~UGetVFromEdgeWithPredOpr() = default;

  std::string get_operator_name() const override {
    return "UGetVFromEdgeWithPredOpr";
  }

  bl::result<Context> Eval(GraphUpdateInterface& graph,
                           const std::map<std::string, std::string>& params,
                           Context&& ctx, OprTimer& timer) override {
    auto expr = parse_expression<GraphUpdateInterface>(
        graph, ctx, params, expr_, VarType::kPathVar);
    if (expr->is_optional()) {
      LOG(ERROR) << "GetV does not support optional expression now";
      RETURN_BAD_REQUEST_ERROR("GetV does not support optional expression now");
    }
    LOG(INFO) << "GetVFromEdgeWithPredOpr";
    return UGetV::get_vertex_from_edge(
        graph, std::move(ctx), params_,
        [&](size_t idx, label_t label, vid_t vid) {
          return expr->eval_vertex(label, vid, idx).as_bool();
        });
  }

 private:
  GetVParams params_;
  common::Expression expr_;
};

class UGetVFromVerticesWithPredOpr : public IUpdateOperator {
 public:
  UGetVFromVerticesWithPredOpr(const GetVParams& params,
                               const common::Expression& expr)
      : params_(params), expr_(expr) {}
  ~UGetVFromVerticesWithPredOpr() = default;

  std::string get_operator_name() const override {
    return "UGetVFromVerticesWithPredOpr";
  }

  bl::result<Context> Eval(GraphUpdateInterface& graph,
                           const std::map<std::string, std::string>& params,
                           Context&& ctx, OprTimer& timer) override {
    auto expr = parse_expression<GraphUpdateInterface>(
        graph, ctx, params, expr_, VarType::kPathVar);
    if (expr->is_optional()) {
      LOG(ERROR) << "GetV does not support optional expression now";
      RETURN_BAD_REQUEST_ERROR("GetV does not support optional expression now");
    }
    return UGetV::get_vertex_from_vertices(
        graph, std::move(ctx), params_,
        [&](size_t idx, label_t label, vid_t vid) {
          return expr->eval_vertex(label, vid, idx).as_bool();
        });
  }

 private:
  GetVParams params_;
  common::Expression expr_;
};
std::unique_ptr<IUpdateOperator> UVertexOprBuilder::Build(
    const Schema& schema, const physical::PhysicalPlan& plan, int op_idx) {
  const auto& vertex = plan.plan(op_idx).opr().vertex();
  int alias = vertex.has_alias() ? vertex.alias().value() : -1;
  int tag = vertex.has_tag() ? vertex.tag().value() : -1;
  GetVParams params;
  VOpt opt = parse_opt(vertex.opt());
  params.opt = opt;
  params.tag = tag;
  params.alias = alias;
  params.tables = parse_tables(vertex.params());
  if (vertex.params().has_predicate()) {
    if (opt == VOpt::kEnd || opt == VOpt::kStart || opt == VOpt::kOther) {
      return std::make_unique<UGetVFromEdgeWithPredOpr>(
          params, vertex.params().predicate());
    } else if (opt == VOpt::kItself) {
      return std::make_unique<UGetVFromVerticesWithPredOpr>(
          params, vertex.params().predicate());
    }
    return std::make_unique<UGetVFromEdgeWithPredOpr>(
        params, vertex.params().predicate());
  }
  if (opt == VOpt::kEnd || opt == VOpt::kStart || opt == VOpt::kOther) {
    return std::make_unique<UGetVFromEdgeOpr>(params);
  }
  LOG(ERROR) << "GetV does not support opt " << static_cast<int>(opt);
  return nullptr;
}
}  // namespace ops
}  // namespace runtime
}  // namespace gs